File: mobile/src/features/platform/api-client/domain-error.ts
/**
 * フロント側 DomainError 型と ProblemDetails → DomainError 変換（ALG-MAP / ERR-01〜08）。
 */

import type { ProblemDetails } from '@yudane/schema';

/** エラーカテゴリ（第1階層、ERR-02）。 */
export type ErrorCategory =
  | 'validation'
  | 'auth'
  | 'not-found'
  | 'conflict'
  | 'safeguard'
  | 'external-api'
  | 'rate-limit'
  | 'internal';

/** フロント側ドメインエラー。 */
export class DomainError extends Error {
  readonly category: ErrorCategory;
  readonly code: string;
  readonly status: number;
  readonly retryable: boolean;
  readonly userMessage: string | undefined;

  constructor(params: {
    category: ErrorCategory;
    code: string;
    message: string;
    status: number;
    retryable?: boolean;
    userMessage?: string;
  }) {
    super(params.message);
    this.name = 'DomainError';
    this.category = params.category;
    this.code = params.code;
    this.status = params.status;
    this.retryable = params.retryable ?? false;
    this.userMessage = params.userMessage;
  }
}

const ERROR_TYPE_BASE = 'https://api.yudane.app/errors/';
const RETRYABLE_STATUS = new Set([429, 503]);

/** HTTP ステータスからカテゴリの既定を導く。 */
function categoryFromStatus(status: number): ErrorCategory {
  if (status === 400 || status === 422) return 'validation';
  if (status === 401) return 'auth';
  if (status === 403) return 'auth';
  if (status === 404) return 'not-found';
  if (status === 409) return 'conflict';
  if (status === 429) return 'rate-limit';
  if (status >= 500) return 'internal';
  return 'internal';
}

/**
 * ProblemDetails を DomainError に変換する（ALG-MAP）。
 *
 * @param problem - サーバーが返した ProblemDetails（パース済み）。null 可
 * @param status - HTTP ステータス
 * @param isIdempotent - 冪等メソッド（GET）か。retryable 判定に使用
 * @returns DomainError
 */
export function mapProblemToDomainError(
  problem: ProblemDetails | null,
  status: number,
  isIdempotent: boolean,
): DomainError {
  const code = problem?.type?.startsWith(ERROR_TYPE_BASE)
    ? problem.type.slice(ERROR_TYPE_BASE.length)
    : `${categoryFromStatus(status)}.unknown`;
  const category = (code.split('.')[0] ?? categoryFromStatus(status)) as ErrorCategory;
  const retryable = RETRYABLE_STATUS.has(status) && isIdempotent;

  return new DomainError({
    category,
    code,
    message: problem?.detail ?? `HTTP ${status}`,
    status,
    retryable,
    userMessage: problem?.title,
  });
}

File: backend/src/common/exceptions/domain_error.py
"""DomainError 体系（domain-entities.md §1 / business-rules.md ERR-01〜08）。

カテゴリ + コードの 2 階層（Q7=A）。API 境界では RFC 7807 ProblemDetails に変換する。
internal カテゴリは detail に内部情報を出さない（SECURITY-09 / ERR-05）。
"""

from __future__ import annotations

from enum import Enum
from typing import TypedDict


class ErrorCategory(str, Enum):
    """エラーカテゴリ（第1階層、ERR-02）。"""

    VALIDATION = "validation"
    AUTH = "auth"
    NOT_FOUND = "not-found"
    CONFLICT = "conflict"
    SAFEGUARD = "safeguard"
    EXTERNAL_API = "external-api"
    RATE_LIMIT = "rate-limit"
    INTERNAL = "internal"


# カテゴリ → 既定 HTTP ステータス（ERR-08）
_DEFAULT_STATUS: dict[ErrorCategory, int] = {
    ErrorCategory.VALIDATION: 400,
    ErrorCategory.AUTH: 401,
    ErrorCategory.NOT_FOUND: 404,
    ErrorCategory.CONFLICT: 409,
    ErrorCategory.SAFEGUARD: 409,
    ErrorCategory.EXTERNAL_API: 502,
    ErrorCategory.RATE_LIMIT: 429,
    ErrorCategory.INTERNAL: 500,
}

_ERROR_TYPE_BASE = "https://api.yudane.app/errors/"


class ProblemDetailsDict(TypedDict, total=False):
    """RFC 7807 ProblemDetails の辞書表現。"""

    type: str
    title: str
    status: int
    detail: str
    instance: str


class DomainError(Exception):
    """ドメインエラー（カテゴリ + コードの 2 階層）。

    Attributes:
        category: 第1階層カテゴリ。
        code: 第2階層コード（`<category>.<slug>` 形式）。
        message: 開発者向け内部メッセージ（ユーザー表示しない）。
        user_message: ユーザー表示用（任意、友達系トーン）。
        status: HTTP ステータス。未指定ならカテゴリ既定値。
        retryable: リトライ可能か（429/503 かつ冪等のみ true）。
    """

    def __init__(
        self,
        category: ErrorCategory,
        code: str,
        message: str,
        *,
        user_message: str | None = None,
        status: int | None = None,
        retryable: bool = False,
        instance: str | None = None,
    ) -> None:
        super().__init__(message)
        self.category = category
        self.code = code
        self.message = message
        self.user_message = user_message
        self.status = status if status is not None else _DEFAULT_STATUS[category]
        self.retryable = retryable
        self.instance = instance


def to_problem_details(error: DomainError) -> ProblemDetailsDict:
    """DomainError を RFC 7807 ProblemDetails に変換する（ERR-03/04/05）。

    internal カテゴリは detail に内部情報を出さず固定文言にする（SECURITY-09）。

    Args:
        error: 変換対象。

    Returns:
        ProblemDetails 辞書（API レスポンス body）。
    """
    problem: ProblemDetailsDict = {
        "type": f"{_ERROR_TYPE_BASE}{error.code}",
        "title": error.user_message or _safe_title(error),
        "status": error.status,
    }
    if error.category is not ErrorCategory.INTERNAL:
        # internal は detail を出さない（ERR-05）
        problem["detail"] = error.message
    if error.instance is not None:
        problem["instance"] = error.instance
    return problem


def _safe_title(error: DomainError) -> str:
    """ユーザー表示用タイトルの既定値（internal は汎用文言）。"""
    if error.category is ErrorCategory.INTERNAL:
        return "予期しないエラーが発生しました"
    return error.code
